linalg/sparsemat.cpp

```diff
@@ -2145,31 +2145,47 @@ void SparseMatrix::DiagScale(const Vector &b, Vector &x, double sc) const
 {
    MFEM_VERIFY(Finalized(), "Matrix must be finalized.");
 
+   const int nnz = J.Capacity();
+
+   const bool use_dev = b.UseDevice() || x.UseDevice();
+
+   auto bp = b.Read(use_dev);
+   auto xp = x.Write(use_dev);
+
+   auto Ap = Read(A, nnz);
+   auto Ip = Read(I, height+1);
+   auto Jp = Read(J, nnz);
+
    bool scale = (sc != 1.0);
-   for (int i = 0, j = 0; i < height; i++)
+   MFEM_FORALL(i, height,
    {
-      int end = I[i+1];
-      for ( ; true; j++)
+      int end = Ip[i+1];
+      for (int j = Ip[i]; true; j++)
       {
-         MFEM_VERIFY(j != end, "Couldn't find diagonal in row. i = " << i
-                     << ", j = " << j
-                     << ", I[i+1] = " << end );
-         if (J[j] == i)
+         if (j == end)
          {
-            MFEM_VERIFY(std::abs(A[j]) > 0.0, "Diagonal " << j << " must be nonzero");
+            printf("Couldn't find diagonal in row. i = %d, j = %d, I[i+1] = %d\n", i, j,
+            end);
```

Comment:
Here we will need something that aborts the kernel, the following macros could be a solution: 
```
#if !defined(NDEBUG)                                                                                                                        
#define MFEM_ABORT_KERNEL assert(0);                                                                                                         
#endif                                                                                                                                      
                                                                                                                                            
#if defined(NDEBUG) && defined(__CUDA_ARCH__)                                                                                               
#define MFEM_ABORT_KERNEL asm("trap;");                                                                                                        
#endif                                                                                                                                      
                                                                                                                                            
#if defined(NDEBUG) && !defined(__CUDA_ARCH__)                                                                                              
#define MFEM_ABORT_KERNEL abort();                                                                                                       
#endif       
```

Reply:
I added this to globals.hpp since it may be useful in other parts of the code.

Reply:
I think you could rewrite `MFEM_VERIFY` instead though, like this we will be able to use it on device.